apps/aecore/src/aec_fork_signalling.erl

```diff
@@ -0,0 +1,350 @@
+        P when ?IS_POINTER(P) ->
+            %% The PrevKeyhash is associated with a pointer that is a block
+            %% hash at height HE - 1, which has the result. The BlockHash is
+            %% set to point to the same block hash at height HE - 1.
+            %% The PrevKeyhash associated with the pointer can now be discarded
+            %% to avoid caching results for all the blocks between HE and
+            %% H2 - 1 (inclusive). Only the block hash at height H2 is
+            %% important.
+            Results1 = add_result(BlockHash, P, Results),
+            Results2 = del_result(PrevKeyHash, Results1),
+            State#state{results = Results2};
+        undefined ->
+            %% There is no result reachable via PrevKeyhash. A new process is
+            %% spawned, it will set both a pointer (for BlockHash) and
+            %% a result for a block hash at height HE - 1.
+            spawn_worker(Block, BlockHash, State)
+    end.
+
+worker_process(Parent, Block, BlockHash, Fork) ->
+    case search_last_signalling_block(Block, BlockHash, Fork) of
+        {ok, LastSigBlock, LastSigBlockHash} ->
+            send_worker_msg(Parent, {check_result, BlockHash, LastSigBlockHash}),
+            case await_server_msg(Parent) of
+                {ok, compute} ->
+                    case compute_signalling_result(LastSigBlock, LastSigBlockHash, Fork) of
+                        {ok, Result} ->
+                            send_worker_msg(Parent, {add_result, LastSigBlockHash, Result});
+                        {error, _Rsn} = Err ->
+                            Err
+                    end;
+                {ok, abort} ->
+                    aborted;
+                {error, _Rsn} = Err ->
+                    Err
+            end;
+        {error, _Rsn} = Err ->
+            Err
+    end.
+
+search_last_signalling_block(Block, BlockHash, Fork) ->
+    case is_last_signalling_block(Block, Fork) of
+        true ->
+            {ok, Block, BlockHash};
+        false ->
+            PrevKeyHash = aec_blocks:prev_key_hash(Block),
+            case get_key_block(PrevKeyHash) of
+                Block1 when Block1 =/= not_found ->
+                    search_last_signalling_block(Block1, PrevKeyHash, Fork);
+                not_found ->
+                    {error, block_not_found}
+            end
+    end.
+
+compute_signalling_result(Block, BlockHash, Fork) ->
+    case compute_signalling_result(Block, BlockHash, Fork, 0) of
+        {ok, Count}         -> signalling_result(Count, Fork);
+        {error, _Rsn} = Err -> Err
+    end.
+
+compute_signalling_result(Block, _BlockHash, Fork, Count) ->
+    case is_first_signalling_block(Block, Fork) of
+        true ->
+            {ok, Count + count_inc(is_matching_info_present(Block, Fork))};
+        false ->
+            PrevKeyHash = aec_blocks:prev_key_hash(Block),
+            case get_key_block(PrevKeyHash) of
```

Comment:
Any reasons why not only the header?

Reply:
Key block is just a header, right? I can use `aec_chain:get_header`, but I assume it will return the same result.

Reply:
You are correct. So it does not matter.